### PR TITLE
[breaking batch] Rename and refactor ast::Pat_ variants

### PR DESCRIPTION
**Pull request by petrochenkov**
cc https://github.com/rust-lang/rust/pull/31487#issuecomment-182945101
plugin-[breaking-change]

The first commit renames `ast::Pat_` to `ast::PatKind` and uses its variants in enum qualified form. I've also taken the opportunity and renamed `PatKind::Region` into `PatKind::Ref`.

The second commit splits `PatKind::Enum` into `PatKind::TupleStruct` and `PatKind::UnitStruct`.
So, pattern kinds now correspond to their struct/variant kinds - `Struct`, `TupleStruct` and `UnitStruct`.
@nikomatsakis @nrc @arielb1 Are you okay with this naming scheme?
An alternative possible naming scheme is `PatKind::StructVariant`, `PatKind::TupleVariant`, `PatKind::UnitVariant` (it's probably closer to the common use, but I like it less).

I intend to apply these changes to HIR later, they should not necessarily go in the same nightly with https://github.com/rust-lang/rust/pull/31487
r? @Manishearth

### CODE REVIEW

**Comment by Manishearth** [comment]
r=me for this, but I'd like someone else to sign off on the naming scheme too.

**Comment by petrochenkov** [comment]
Rebased.

**Comment by Manishearth** [comment]
@nrc @eddyb can you approve the change with variants here?

File: src/libsyntax/ast.rs [hunk 1]
```diff
@@ -552,47 +551,54 @@ pub enum BindingMode {
...
+
+    /// A tuple struct/variant pattern `Variant(x, y, z)`.
     /// "None" means a `Variant(..)` pattern where we don't bind the fields to names.
-    PatEnum(Path, Option<Vec<P<Pat>>>),
+    TupleStruct(Path, Option<Vec<P<Pat>>>),
+
+    /// A unit struct/variant pattern.
+    /// Some const patterns are also represented by `PatKind::UnitStruct` due to parser limitations.
+    /// Some unit patterns are represented by `PatKind::Ident` due to the same parser limitations.
+    UnitStruct(Path),
```
Comment (eddyb) [inline src/libsyntax/ast.rs]: I think this should be just `Path`.

**Comment by eddyb** [comment]
LGTM except for the `UnitStruct` name.

File: src/libsyntax/ast.rs (hunk 1 above)
Comment (petrochenkov) [inline src/libsyntax/ast.rs]: >I think this should be just `Path`.

Yeah, probably, for AST and for HIR for the time being.
I'd like to have something like this in the HIR eventually, when name resolution is performed before lowering to HIR:
pat

Comment (eddyb) [inline src/libsyntax/ast.rs]: Associated constants and top-level constants are different though.

**Comment by petrochenkov** [comment]
Updated.

**Comment by petrochenkov** [comment]
Ping @Manishearth

**Comment by Manishearth** [comment]
@bors r+ 

cc @erickt

**Pull request merged**